unumpy/_multimethods.py

```diff
@@ -1156,3 +1156,21 @@ def _allclose_default(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
 @all_of_type(ndarray)
 def allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
     return a, b
+
+
+def diag_default(v, k=0):
+    if ndim(v) == 1:
+        a = zeros((len(v), len(v)), dtype=v.dtype)
+        for i in range(len(v)):
+            a[i, i] = v[i]
```

Comment:
I haven't included the test for `diag`. The Sparse backend fails in this line with the following error:

> TypeError: 'COO' object does not support item assignment

Reply:
I would avoid defaults with long loops. :)

Reply:
How can we return the diagonal matrix without using a loop and item assignment then?

Reply:
In general, we'd avoid even having defaults with long loops or assignment. We'd just not have a default in that case.

Reply:
Not sure it's relevant in this case, but the diagonal is just a strided view of the flattened array:
```python
In [7]: N = 5

In [8]: A = np.arange(N*N).reshape(N, N)

In [9]: np.diag(A)
Out[9]: array([ 0,  6, 12, 18, 24])

In [10]: A.ravel()[0::(N+1)]
Out[10]: array([ 0,  6, 12, 18, 24])

In [11]: A.ravel()[0::(N+1)] = 9

In [12]: A
Out[12]:
array([[ 9,  1,  2,  3,  4],
       [ 5,  9,  7,  8,  9],
       [10, 11,  9, 13, 14],
       [15, 16, 17,  9, 19],
       [20, 21, 22, 23,  9]])
```

Reply:
Yes, extremely relevant in fact. This helped me reimplement `diag_default`.